fyrox-impl/src/scene/tilemap/effect.rs

```diff
@@ -0,0 +1,263 @@
+        if let (Some(material), Some(offset)) = (self.material.as_ref(), self.offset) {
+            for &position in self.positions.iter() {
+                let position = position + offset;
+                push_highlight(position, material, self.thickness, context);
+            }
+        }
+    }
+}
+
+/// Sets the tiles at the given positions to invisible.
+#[derive(Debug)]
+pub struct TileEraseEffect {
+    /// The positions of the tiles to hide
+    pub positions: FxHashSet<Vector2<i32>>,
+}
+
+impl TileMapEffect for TileEraseEffect {
+    fn render_special_tiles(&self, context: &mut TileMapRenderContext) {
+        for &position in self.positions.iter() {
+            context.set_tile_visible(position, false);
+        }
+    }
+}
+
+/// Draws the given tiles with the given offset, and sets the drawn tile positions
+/// to invisible so that no other tiles will be drawn there.
+#[derive(Debug)]
+pub struct TileOverlayEffect {
+    /// True if the tiles are to be drawn. If false, then this effect does nothing.
+    pub active: bool,
+    /// Vector to be added to the positions of the tiles before rendering
+    pub offset: Vector2<i32>,
+    /// The tiles to render
+    pub tiles: FxHashMap<Vector2<i32>, TileDefinitionHandle>,
+}
+
+impl TileMapEffect for TileOverlayEffect {
+    fn render_special_tiles(&self, context: &mut TileMapRenderContext) {
+        if !self.active {
+            return;
+        }
+        for (&position, &handle) in self.tiles.iter() {
+            let position = position + self.offset;
+            if context.is_tile_visible(position) {
+                context.draw_tile(position, handle);
+                context.set_tile_visible(position, false);
+            }
+        }
+    }
+}
+
+/// Uses the given tile update to render the replacement tiles and make
+/// the erased tiles invisible.
+#[derive(Debug)]
+pub struct TileUpdateEffect {
+    /// True if the tiles are to be drawn. If false, then this effect does nothing.
+    pub active: bool,
+    /// The update data to render in the tile map
+    pub update: TransTilesUpdate,
+}
+
+impl TileMapEffect for TileUpdateEffect {
+    fn render_special_tiles(&self, context: &mut TileMapRenderContext) {
+        if !self.active {
+            return;
+        }
+        for (&position, value) in self.update.iter() {
+            if context.is_tile_visible(position) {
+                context.set_tile_visible(position, false);
+                if let Some((transform, handle)) = *value {
+                    let handle = context
+                        .tile_set
+                        .get_transformed_version(transform, handle)
+                        .unwrap_or(handle);
+                    context.draw_tile(position, handle);
+                }
+            }
+        }
+    }
+}
+
+fn make_highlight_vertex(transform: &Matrix4<f32>, position: Vector2<f32>) -> StaticVertex {
+    StaticVertex {

```

Comment:
Ideally, tile map should use specialized vertex format to reduce memory usage. This vertex format is specific for 3d, because it contains normal and tangent (which are non-existent in 2d).